### ISSUE TITLE
ARea UPdate system causes major performance overhead

### ISSUE THREAD

**Issue by Crystalwarrior** [AttorneyOnline/AO2-Client#241]
Despite the fact that this system was conceptualized precisely to *reduce* performance by not having a need to regenerate the whole area list but instead just update a few numbers, that's precisely what ends up happening anyway - I'm presuming each ARUP packet regenerates the whole area list, and if the server decided to update a lot of information at once, the client would end up chugging along trying to regenerate the same list 4 times in quick succession.

This problem is two-fold:
1) there is no way to combine information sent by ARUP - each ARUP packet has to be sent separately, so you can't update CMs and Players at once without performance impact.
2) disconnect packet causes an ARUP to be called, and this occurs with -1 characters aka the MS Probe. That means every time the MS probe joins and disconnects, ARUP packets are called, and that ends up bombaring the client with constant lag spikes. This can be optimized on the serverside, of course, however the mere fact that calling something as simple as "update number of players for this area from 0 to 1" causes *this much* performance impact is a severe problem, as the server sending its own ARUP style system with regenerating the whole list would be *more optimized*.

**Comment by Crystalwarrior** [comment]
Here's an example, every time the server needs to be regenerated with new ARUP info for whatever reason, this is the funcs called:
```
area_manager.send_arup_players()
area_manager.send_arup_status()
area_manager.send_arup_cms()
area_manager.send_arup_lock()
```
This causes everyone's Area List to be regenerated **FOUR TIMES**, and there is no way to send all this information in a single packet.

**Comment by gameboyprinter** [comment]
https://github.com/AttorneyOnline/AO2-Client/blob/master/src/courtroom.cpp#L1431
This function in particular should be optimized, and yes, the area updates should only regenerate the area list once.

**Comment by oldmud0** [comment]
Likely I/O-bound by multiple calls to `ao_app->get_color`.

**Comment by gameboyprinter** [comment]
couldn't that be cached into memory earlier on?

**Comment by oldmud0** [comment]
That is correct.

**Comment by scatterflower** [comment]
Ok, so, I looked more carefully at that code snippet. It wasn't regenerating four times; see, the update_list() was called every single time the code inside the for loop ran. So, if you updated, say, 5 areas, then update_list() would be called 5 times. On top of this, IO was being performed multiple times in update_list(). With this PR, the IO only happens when the courtroom is created, and update_list() is only called once per ARUP.

**Comment by scatterflower** [comment]
Fixed by #244
